### ISSUE TITLE
[no-void]: allow void arrow functions

### ISSUE THREAD

**Issue by G-Rath** [eslint/eslint#13299]
**What rule do you want to change?**
`no-void`

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
Option - could also be implemented as standard behaviour depending on interpretation of the rules description.

**Please provide some example code that this change will affect:**

```js
const log = x => void console.log(x);
```

**What does the rule currently do for this code?**
Errors

**What will the rule do after it's changed?**
Not error

**Are you willing to submit a pull request to implement this change?**
Yes, I have a PR open: #13268

-----

There are some [legitimate uses](https://gist.github.com/slikts/dee3702357765dda3d484d8888d3029e) for the void operator, one of which is even mentioned in the opening sentence of the `no-void` docs:

>  It can be used to ignore any side effects expression may produce

The description of the `no-void` rule to me seems to indicate that the rule is actually ideally about preventing the use of `void` to get a "pure" undefined back in the pre-ES5 days when undefined was mutable in global scope: 

> The common case of using void operator is to get a "pure" undefined value as prior to ES5 the undefined variable was mutable:

So it might be arguable that this is technically a bug? But I don't have a major preference on if it's an option or not :)

Based off #12688, which was closed for not following the template.
Based off #13020, which didn't get much support, but I still think this is a useful minor addition.

**Comment by kaicataldo** [comment]
@eslint/eslint-team Can we get a few more opinions about this proposal? There is an open PR for this change.

**Comment by nzakas** [comment]
This doesn’t seem like a new option, it seems [allowAsStatement](https://eslint.org/docs/rules/no-void#options) should allow an arrow function with void in the return value. If that’s not the case, then I’d say this is a bug.

**Comment by ljharb** [comment]
I don't see why; `x => void console.log(x)` does not have `void` in a statement position, but in an expression position. In other words, the "allow as statement" option should not impact expressions in any way.

**Comment by nzakas** [comment]
Oops sorry, read the docs too quickly and got confused. You’re right, my previous comment makes no sense; please disregard.

That said, I’m not in favor of adding more options to this rule. We are intentionally no longer adding exceptions to stylistic rules, so unless someone wants to argue that this is a bug, then I think a custom rule is the better solution.

**Comment by G-Rath** [comment]
I don't mind arguing that this is a bug. I've always felt the primary target of the rule was on using `void` to get a "pure" `undefined` (which is now considered legacy), and attempts to reduce code size via `void 0` (which is considered hard to read, and largely unneeded).

Meanwhile, I believe arrow functions didn't land until about a year after the `no-void` rule was released, meaning this usage probably wouldn't have been on the author's mind.

Additionally as typed languages such as TypeScript become more and more prevalent, the places where `void` shines as a readable & useful operator are starting to.. well.. shine through 😄

-----

I think it's a sensible choice for ESLint core to start pushing back against stylish rules, but don't think that the existing ones should be frozen unless deprecated, as I feel that the ESLint docs are considered a Source of Truth on if something is good practice or not (regardless of if that's what they're meant to be or not).

On that note, I'd be happy setup & help look after an eslint-plugin to allow these stylistic rules to be deprecated in the core in favor of said plugin (something like `eslint-plugin-stylistic`?)

**Comment by mdjermanovic** [comment]
> Additionally as typed languages such as TypeScript become more and more prevalent, the places where `void` shines as a readable & useful operator are starting to.. well.. shine through

Are there some places other than arrow function's implicit return?

**Comment by btmills** [comment]
The docs for the [`allowAsStatement` option](https://eslint.org/docs/rules/no-void#allowasstatement) don't suggest a use case for the option, but the [pull request that added it](https://github.com/eslint/eslint/pull/12613) cited [`@typescript-eslint/no-floating-promises`](https://github.com/typescript-eslint/typescript-eslint/blob/00ac9c3ccaad27bab08ec3c3a104f612bb593df5/packages/eslint-plugin/docs/rules/no-floating-promises.md), where using `void someFunctionCall();` indicates "yes, I'm aware this function returns a value, and I'm showing explicitly that I don't need it." From that perspective, `() => void someFunctionCall()` is fulfilling the same purpose of explicitly rather than implicitly discarding a return value. One could argue that `allowAsStatement` should apply to statement*ish* `void`ed arrow function expression bodies.

What do you all think? Is that a cogent argument or just mental gymnastics?

**Comment by G-Rath** [comment]
> Are there some places other than arrow function's implicit return?

`allowAsStatements` is at least one more.

>  One could argue that allowAsStatement should apply to statementish voided arrow function expression bodies.

I'm also happy to go with that option - while as @ljharb pointed out, the expression is not a statement, I think the concept of having these two behaviours controlled by the same flag is not unreasonable.

**Comment by ljharb** [comment]
It seems much easier tho to add a new option rather than confusing the meaning of an existing one.

**Comment by nzakas** [comment]
I'd like the team to focus on answering this question: do we want to make a change here or not? So far, there is little stated support on the team and we seem to be discussing whether an existing solution works or if a new one is needed.

So, if you'd like to make a change to support this case, please state that. If enough people want to make a change, then I think it makes sense to discuss what type of change to make. 

As stated earlier, I'm against making a change for this.

**Comment by G-Rath** [comment]
@yeonjuan [seemed to in support of it](https://github.com/eslint/eslint/issues/13020#issuecomment-596467520):

> @G-Rath
> Thanks for the proposal. 👍
> I think it makes sense. I labeled it evaluating to see what other team members think.

(cross-posting their comment from the previous ticket, as they might not have seen this new one - happy to amend this comment if their feelings have changed since then 🙂)

**Comment by v4dkou** [comment]
The reasoning behind the `no-void` was never to disallow non-leaking arrow functions.
https://github.com/eslint/eslint/issues/1017

The rule was created in 2014, while ES6 Arrow functions were still in proposal stage and were only published in 2015.

At the time, the current implementation was the simplest for the intended purpose, and nobody bothered to update the old rule since ES6 has finalized.

I would argue, the current implementation is outdated and should be improved by not blocking void in return values for arrow functions _by default_, without adding options for stylistic choices.

**Comment by mdjermanovic** [comment]
If there's intention to use `void` and it's seen as a useful language feature, then why not just turn the rule off, as stated in the [documentation](https://eslint.org/docs/rules/no-void#when-not-to-use-it):

> When Not To Use It
>
> If you intentionally use the void operator then you can disable this rule.

**Comment by v4dkou** [comment]
@mdjermanovic 
> If there's intention to use `void` and it's seen as a useful language feature, then why not just turn the rule off

Because some may not consider it to be a _generally_ useful feature, and would prefer it being disallowed for the cases this rule was introduced as per the documentation.

> The common case of using void operator is to get a "pure" undefined value as prior to ES5 the undefined variable was mutable
> ...
> Another common case is to minify code as `void 0` is shorter than `undefined`
> ...
> When used with IIFE (immediately-invoked function expression), void can be used to force the function keyword to be treated as an expression instead of a declaration:

Personally, I agree that the stated cases are obscure, and would like if ESLint could continue catching them automatically, rather than disabling the rule and enforcing the proper usage through code reviews. 

However, there are modern use cases that nobody could've predicted because the language feature they rely upon is more recent than the implementation of the checker. The argument is that the rule checks should evolve together with the language it checks.

In other words, current behaviour seems like an unintended side effect, rather than a deliberate choice.

**Comment by nzakas** [comment]
> However, there are modern use cases that nobody could've predicted because the language feature they rely upon is more recent than the implementation of the checker. The argument is that the rule checks should evolve together with the language they check.

Yes, this is my rationale for saying I don't think it's worth updating the rule. This is a stylistic rule, and I don't think it's worth the overhead to keep modifying it for every new use case that comes along.

**Comment by ljharb** [comment]
How is it stylistic? It doesn't control how the code is formatted (how it looks), it prevents certain syntax from being used.

**Comment by mdjermanovic** [comment]
In case that this option doesn't get accepted, the same functionality can be achieved with `no-restricted-syntax`:

```js
/*eslint no-restricted-syntax: ["error", {
    "selector": ":not(ArrowFunctionExpression) > UnaryExpression[operator='void']",
    "message": "Don't use 'void'."
}]*/

void 0; // error

void function(){}(); // error

const log = x => void console.log(x); // ok
```

[demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVzdHJpY3RlZC1zeW50YXg6IFtcImVycm9yXCIsIHtcbiAgICBcInNlbGVjdG9yXCI6IFwiOm5vdChBcnJvd0Z1bmN0aW9uRXhwcmVzc2lvbikgPiBVbmFyeUV4cHJlc3Npb25bb3BlcmF0b3I9J3ZvaWQnXVwiLFxuICAgIFwibWVzc2FnZVwiOiBcIkRvbid0IHVzZSAndm9pZCcuXCJcbn1dKi9cblxudm9pZCAwOyAvLyBlcnJvclxuXG52b2lkIGZ1bmN0aW9uKCl7fSgpOyAvLyBlcnJvclxuXG5jb25zdCBsb2cgPSB4ID0+IHZvaWQgY29uc29sZS5sb2coeCk7IC8vIG9rIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

**Comment by mdjermanovic** [comment]
"Stylistic" as it's about preference. It doesn't enforce best practice or disallow a bad practice.

This rule is currently in the Best Practices category, but by reading #1017 I believe the intended category was Stylistic Issues.

If `void 0` or any other usage of `void` (aside from arrow functions) is actually considered to be a bad practice, and `x => void console.log(x)` is considered to be okay, I'd vote for this option.

**Comment by nzakas** [comment]
**TSC Summary:** The proposal here is to add a new option to `no-void` that will allow the form `() => void foo;`, which currently emits a warning. Our policy says [we have frozen stylistic rules](https://eslint.org/blog/2020/05/changes-to-rules-policies#what-are-the-changes) and there are other ways to accomplish the same thing (creating a custom rule, using `no-restricted-syntax`, etc.).

**TSC Question:** Should we accept this proposal?

**Comment by btmills** [comment]
Thank you all for the discussion in this issue, and particularly to @G-Rath for implementing the pull request! We talked about this in today's TSC meeting, and the team felt that given the varying opinions and practices around this rule, its current classification as a Best Practice could be confusing. It fits better as a Stylistic rule, which seems to be more consistent with the original intent when it was added. We'll track that change in a separate issue. That also means we're not going to add the new option into the core rule, so please check out @mdjermanovic's [`no-restricted-syntax` example](https://github.com/eslint/eslint/issues/13299#issuecomment-656335560) to achieve the same behavior. @G-Rath if you (or anyone else) release your implementation as a shareable rule in a plugin, please share that here!

**Comment by Chamion** [comment]
I'm disappointed to see this change denied but I understand the reasoning.

I wrote the plugin [eslint-plugin-void-only-side-effects](https://www.npmjs.com/package/eslint-plugin-void-only-side-effects) before I read this thread. While it is more generic than this change (and therefore doesn't serve as a direct replacement) it more fully addresses the concerns of the [article](https://gist.github.com/slikts/dee3702357765dda3d484d8888d3029e) linked by @G-Rath in the opening post.

It allows voiding the return value of a side effect like `void sideEffect()` or `void (ref.current = 1)` but also disallows non-side-effect expressions like `void 0`.

I had hoped for my rule to be included as an option of `no-void` but judging by this conversation I shouldn't bother asking. I hope it's helpful to someone as a plugin 😄.
